tests/src/buildApplication/generateFunctionsMap.test.ts

```diff
@@ -1,148 +1,159 @@
+			path: {
+				'(group-1)': {
+					to: {
+						'(group-2)': {
+							'page.func': validFuncDir,
+						},
+					},
+				},
+			},
+		});
 
-		expect(functionsMap.size).toEqual(10);
-		// index
-		expect(functionsMap.get('/')).toMatch(/\/index\.func\.js$/);
-		expect(functionsMap.get('/index')).toMatch(/\/index\.func\.js$/);
-		expect(functionsMap.get('/index.rsc')).toMatch(/\/index\.rsc\.func\.js$/);
-		// nested route
-		expect(functionsMap.get('/api/hello')).toMatch(/\/api\/hello\.func\.js$/);
-		// middleware
-		expect(functionsMap.get('/middlewarejs')).toMatch(
-			/\/middlewarejs\.func\.js$/
-		);
-		expect(functionsMap.get('/base/middleware')).toMatch(
-			/\/base\/middleware\.func\.js$/
-		);
-		// route group
+		expect(functionsMap.size).toEqual(1);
 		expect(functionsMap.get('/path/to/page')).toMatch(
 			/\/path\/\(group-1\)\/to\/\(group-2\)\/page\.func\.js$/
 		);
+	});
+
+	test('should squash invalid root functions', async () => {
+		const { invalidFunctions, functionsMap } = await generateFunctionsMapFrom({
+			'should-be-valid.func': invalidFuncDir,
+			'(is-actually-valid)': {
+				'should-be-valid.func': validFuncDir,
+			},
+		});
+
+		expect(Array.from(invalidFunctions.values())).toEqual([]);
+		expect(functionsMap.size).toEqual(1);
 		expect(functionsMap.get('/should-be-valid')).toMatch(
-			/\(is-valid\)\/should-be-valid\.func\.js$/
-		);
-		expect(functionsMap.get('/rsc/should-be-valid')).toMatch(
-			/rsc\/\(is-valid\)\/should-be-valid\.func\.js$/
-		);
-		expect(functionsMap.get('/rsc/should-be-valid.rsc')).toMatch(
-			/rsc\/\(is-valid\)\/should-be-valid\.func\.js$/
+			/\(is-actually-valid\)\/should-be-valid\.func\.js$/
 		);
 	});
 
-	// TODO: add tests that also test the functions map with the experimentalMinify flag
-
 	test('should return invalid functions', async () => {
-		const { invalidFunctions } = await generateFunctionsMap(
-			'invalidTest/functions',
-			false
-		);
+		const { invalidFunctions } = await generateFunctionsMapFrom({
+			'index.func': invalidFuncDir,
+			'index.rsc.func': invalidFuncDir,
+		});
 
-		expect(invalidFunctions.size).toEqual(3);
 		expect(Array.from(invalidFunctions.values())).toEqual([
 			'index.func',
 			'index.rsc.func',
-			'should-be-valid-alt.func',
 		]);
 	});
+
+	test('should ignore a generated middleware.js file while also proving a warning', async () => {
+		const fn = vi.fn();
+		// eslint-disable-next-line no-console
+		console.warn = fn;
+
+		const { invalidFunctions } = await generateFunctionsMapFrom({
+			'middlewarejs.func': {
+				'.vc-config.json': JSON.stringify({
+					name: 'middleware',
+					runtime: 'edge',
+					entrypoint: 'middleware.js',
+				}),
+				'middleware.js': '',
+			},
+		});
+
+		expect(Array.from(invalidFunctions.values())).toEqual([]);
+		expect(fn.mock.calls.length).toBe(1);
+		expect(fn.mock.calls[0]?.[0]).toMatch(
+			/invalid middleware function for middlewarejs.func/
+		);
```

Comment:
We should spy on the method and mock its implementation for `console.warn`, instead of setting it to `vi.fn()`. This also means that we can then clean it up and restore the original function at the end of the test.

```suggestion
		const mockedWarn = vi.spyOn(console, 'warn').mockImplementation(() => null);

		const { invalidFunctions } = await generateFunctionsMapFrom({
			'middlewarejs.func': {
				'.vc-config.json': JSON.stringify({
					name: 'middleware',
					runtime: 'edge',
					entrypoint: 'middleware.js',
				}),
				'middleware.js': '',
			},
		});

		expect(Array.from(invalidFunctions.values())).toEqual([]);
		expect(mockedWarn).toHaveBeenCalledTimes(1);
		expect(mockedWarn).toHaveBeenLastCalledWith(
			expect.stringMatching(/invalid middleware function for middlewarejs.func/)
		);

		mockedWarn.mockRestore();
```

Reply:
doh, yeah you're totally right! silly mistake on my part! 😓👍